src/builder.rs

```diff
@@ -27,9 +27,61 @@ use crate::{
 
 const MIN_ACTIONS: usize = 2;
 
+/// An enumeration of rules Orchard bundle construction.
+#[derive(Clone, Copy, Debug, PartialEq, Eq)]
+pub enum BundleType {
+    /// A transactional bundle will be padded if necessary to contain at least 2 actions,
+    /// irrespective of whether any genuine actions are required.
+    Transactional(Flags, Anchor),
+    /// A coinbase bundle is required to have no non-dummy spends. No padding is performed.
+    Coinbase,
+}
+
+impl BundleType {
+    /// Returns the number of logical actions that builder will produce in constructing a bundle
+    /// of this type, given the specified numbers of spends and outputs.
+    ///
+    /// Returns an error if the specified number of spends and outputs is incompatible with
+    /// this bundle type.
+    pub fn num_actions(
+        &self,
+        num_spends: usize,
+        num_outputs: usize,
+    ) -> Result<usize, &'static str> {
+        let num_requested_actions = core::cmp::max(num_spends, num_outputs);
+
+        match self {
+            BundleType::Transactional(flags, _) => {
+                if !flags.spends_enabled() && num_spends > 0 {
+                    Err("Spends are disabled, so num_spends must be zero")
+                } else if !flags.outputs_enabled() && num_outputs > 0 {
+                    Err("Outputs are disabled, so num_outputs must be zero")
+                } else {
+                    Ok(core::cmp::max(num_requested_actions, MIN_ACTIONS))
+                }
+            }
+            BundleType::Coinbase => Ok(num_outputs),
```

Comment:
In looking at this again, I think it's better to make the result of `bundle` a `Result<Option<...>, ...>` to avoid needing to provide an error case here.